File: graph/BidirectionalBFS/java/BidirectionalBFS.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class BidirectionalBFS {
    private BidirectionalBFS() {}

    public static final class Result {
        public final int distance;
        public final List<Integer> path;

        public Result(int distance, List<Integer> path) {
            this.distance = distance;
            this.path = path;
        }
    }

    public static Result shortestPath(List<List<Integer>> graph, int source, int target) {
        if (source == target) {
            return new Result(0, List.of(source));
        }

        int n = graph.size();
        int[] parentFromSource = new int[n];
        int[] parentFromTarget = new int[n];
        for (int i = 0; i < n; i++) {
            parentFromSource[i] = -1;
            parentFromTarget[i] = -1;
        }

        ArrayDeque<Integer> sourceFrontier = new ArrayDeque<>();
        ArrayDeque<Integer> targetFrontier = new ArrayDeque<>();
        parentFromSource[source] = source;
        parentFromTarget[target] = target;
        sourceFrontier.addLast(source);
        targetFrontier.addLast(target);

        while (!sourceFrontier.isEmpty() && !targetFrontier.isEmpty()) {
            int meeting;
            if (sourceFrontier.size() <= targetFrontier.size()) {
                meeting = expandOneLevel(sourceFrontier, parentFromSource, parentFromTarget, graph);
            } else {
                meeting = expandOneLevel(targetFrontier, parentFromTarget, parentFromSource, graph);
            }

            if (meeting != -1) {
                List<Integer> path = buildPath(meeting, parentFromSource, parentFromTarget);
                return new Result(path.size() - 1, path);
            }
        }

        return new Result(-1, List.of());
    }

    private static int expandOneLevel(
        ArrayDeque<Integer> frontier,
        int[] parentThisSide,
        int[] parentOtherSide,
        List<List<Integer>> graph
    ) {
        int levelSize = frontier.size();
        for (int i = 0; i < levelSize; i++) {
            int current = frontier.removeFirst();
            for (int neighbor : graph.get(current)) {
                if (parentThisSide[neighbor] != -1) {
                    continue;
                }

                parentThisSide[neighbor] = current;
                if (parentOtherSide[neighbor] != -1) {
                    return neighbor;
                }
                frontier.addLast(neighbor);
            }
        }
        return -1;
    }

    private static List<Integer> buildPath(int meeting, int[] parentFromSource, int[] parentFromTarget) {
        ArrayList<Integer> left = new ArrayList<>();
        for (int vertex = meeting; vertex != -1; ) {
            left.add(vertex);
            if (parentFromSource[vertex] == vertex) {
                break;
            }
            vertex = parentFromSource[vertex];
        }
        Collections.reverse(left);

        ArrayList<Integer> path = new ArrayList<>(left);
        for (int vertex = parentFromTarget[meeting]; vertex != -1; ) {
            path.add(vertex);
            if (parentFromTarget[vertex] == vertex) {
                break;
            }
            vertex = parentFromTarget[vertex];
        }
        return path;
    }
}

File: graph/BidirectionalBFS/python/BidirectionalBFS.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class BidirectionalBFSResult:
    distance: int
    path: list[int]


def _build_path(
    meeting: int,
    parent_from_source: Sequence[int],
    parent_from_target: Sequence[int],
) -> list[int]:
    left: list[int] = []
    vertex = meeting
    while vertex != -1:
        left.append(vertex)
        if parent_from_source[vertex] == vertex:
            break
        vertex = parent_from_source[vertex]
    left.reverse()

    path = left[:]
    vertex = parent_from_target[meeting]
    while vertex != -1 and vertex != meeting:
        path.append(vertex)
        if parent_from_target[vertex] == vertex:
            break
        vertex = parent_from_target[vertex]
    return path


def _expand_one_level(
    frontier: deque[int],
    parent_this_side: list[int],
    parent_other_side: Sequence[int],
    graph: Sequence[Sequence[int]],
) -> int:
    for _ in range(len(frontier)):
        current = frontier.popleft()
        for neighbor in graph[current]:
            if parent_this_side[neighbor] != -1:
                continue

            parent_this_side[neighbor] = current
            if parent_other_side[neighbor] != -1:
                return neighbor
            frontier.append(neighbor)
    return -1


def shortest_path_bidirectional_bfs(
    graph: Sequence[Sequence[int]],
    source: int,
    target: int,
) -> BidirectionalBFSResult:
    if source == target:
        return BidirectionalBFSResult(distance=0, path=[source])

    n = len(graph)
    parent_from_source = [-1] * n
    parent_from_target = [-1] * n
    source_frontier: deque[int] = deque([source])
    target_frontier: deque[int] = deque([target])

    parent_from_source[source] = source
    parent_from_target[target] = target

    while source_frontier and target_frontier:
        if len(source_frontier) <= len(target_frontier):
            meeting = _expand_one_level(
                source_frontier,
                parent_from_source,
                parent_from_target,
                graph,
            )
        else:
            meeting = _expand_one_level(
                target_frontier,
                parent_from_target,
                parent_from_source,
                graph,
            )

        if meeting != -1:
            path = _build_path(meeting, parent_from_source, parent_from_target)
            return BidirectionalBFSResult(distance=len(path) - 1, path=path)

    return BidirectionalBFSResult(distance=-1, path=[])
